platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -461,8 +460,8 @@ public void add(SingletonService serviceProvider) {
             assertMutable();
             stoppable.add(serviceProvider);
             collectProvidersForClassHierarchy(inspector, serviceProvider.getDeclaredServiceTypes(), serviceProvider);
-            services.add(serviceProvider);
-            for (AnnotatedServiceLifecycleHandler annotationHandler : lifecycleHandlers) {
+            ServicesSnapshot snapshot = services.updateAndGet(it -> it.addService(serviceProvider));
+            for (AnnotatedServiceLifecycleHandler annotationHandler : snapshot.lifecycleHandlers) {
```

Comment:
```suggestion
            for (AnnotatedServiceLifecycleHandler annotationHandler : snapshot.services.updateAndGet(it -> it.addService(serviceProvider))) {
```
could inline to keep high cohesion promise reduce coupling context. Considering snapshot service will probably provide some snapshot.